### PR TITLE
Ignore incomplete structs/classes completely in C++/CLI

### PR DESCRIPTION
**Pull request by realvictorprm**

### CODE REVIEW

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 1]
```diff
@@ -636,6 +636,11 @@ public void GenerateClassProperties(Class @class)
             PopIndent();
         }
 
+        internal static bool PropertyIgnored(Property prop)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: This method should be inlined.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 2]
```diff
@@ -806,6 +811,23 @@ public void GenerateFunction(Function function)
             PopBlock();
         }
 
+        internal static bool FunctionIgnored(Function function)
+        {
+            foreach (var param in function.Parameters)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Simplify by using LINQ and ||.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 3]
```diff
@@ -806,6 +811,23 @@ public void GenerateFunction(Function function)
...
+                if (TypeIgnored(param.Type)) return true;
+            return TypeIgnored(function.ReturnType.Type);
+        }
+
+        internal static bool TypeIgnored(CppSharp.AST.Type type)
+        {
+            var desugared = type.Desugar();
+            var finalType = (desugared.GetFinalPointee() ?? desugared).Desugar();
+            var @class = new Class();
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Don't create a class, just:

Class @class;

is enough.

File: src/Generator/Generators/CLI/CLISources.cs [hunk 4]
```diff
@@ -686,7 +687,7 @@ private void GenerateStructMarshaling(Class @class, string nativeVar)
             }
 
             int paramIndex = 0;
-            foreach (var property in @class.Properties.Where( p => !ASTUtils.CheckIgnoreProperty(p)))
+            foreach (var property in @class.Properties.Where( p => !ASTUtils.CheckIgnoreProperty(p) && !CLIHeaders.PropertyIgnored(p)))
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLISources.cs]: Move the last condition on a new line.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 5]
```diff
@@ -806,6 +808,22 @@ public void GenerateFunction(Function function)
             PopBlock();
         }
 
+        internal static bool FunctionIgnored(Function function)
+        {
+            return TypeIgnored(function.ReturnType.Type) || 
+                   function.Parameters.Any(param => TypeIgnored(param.Type));
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Decrease the indentation to 4 spaces.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 6]
```diff
@@ -806,6 +808,22 @@ public void GenerateFunction(Function function)
             PopBlock();
         }
 
+        internal static bool FunctionIgnored(Function function)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Make private, or public if you use it externally.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 7]
```diff
@@ -806,6 +808,22 @@ public void GenerateFunction(Function function)
...
+        {
+            return TypeIgnored(function.ReturnType.Type) || 
+                   function.Parameters.Any(param => TypeIgnored(param.Type));
+        }
+
+        internal static bool TypeIgnored(CppSharp.AST.Type type)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Make private, or public if you use it externally.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 8]
```diff
@@ -806,6 +808,22 @@ public void GenerateFunction(Function function)
...
+        {
+            var desugared = type.Desugar();
+            var finalType = (desugared.GetFinalPointee() ?? desugared).Desugar();
+            Class @class;
+            if (finalType.TryGetClass(out @class) &&
+                @class.IsIncomplete) return true;
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Move @class.IsIncomplete to the previous line.

File: src/Generator/Generators/CLI/CLISources.cs [hunk 9]
```diff
@@ -742,6 +744,7 @@ private bool GenerateClassConstructorBase(Class @class, Method method = null)
 
         public void GenerateMethod(Method method, Class @class)
         {
+            if (CLIHeaders.FunctionIgnored(method)) return;
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLISources.cs]: Move the return to a new line.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 10]
```diff
@@ -636,6 +636,8 @@ public void GenerateClassProperties(Class @class)
             PopIndent();
         }
 
+        internal static bool PropertyIgnored(Property prop){ return TypeIgnored(prop.Type); }
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: By inlining I meant - remove this method and replace its usages with its body.

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 11]
```diff
@@ -622,7 +622,7 @@ public void GenerateClassProperties(Class @class)
             }
 
             PushIndent();
-            foreach (var prop in @class.Properties.Where(prop => !ASTUtils.CheckIgnoreProperty(prop)))
+            foreach (var prop in @class.Properties.Where(prop => !ASTUtils.CheckIgnoreProperty(prop) && !PropertyIgnored(prop)))
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLIHeaders.cs]: Move the last condition to a new line.

File: src/Generator/Generators/CLI/CLISources.cs [hunk 12]
```diff
@@ -237,7 +237,8 @@ private void GenerateClassProperties(Class @class, Class realOwner)
             }
 
             foreach (var property in @class.Properties.Where(
-                p => !ASTUtils.CheckIgnoreProperty(p) && !p.IsInRefTypeAndBackedByValueClassField()))
+                p => !ASTUtils.CheckIgnoreProperty(p) && !p.IsInRefTypeAndBackedByValueClassField() &&
+                     !CLIHeaders.PropertyIgnored(p)))
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLISources.cs]: Indendation - 4 spaces.

File: tests/CLI/CLI.h [hunk 13]
```diff
@@ -48,4 +48,11 @@ std::ostream& operator<<(std::ostream& os, const Date& dt)
     return os;
 }
 
-DLL_API void testFreeFunction();
\ No newline at end of file
+DLL_API void testFreeFunction();
+
+struct CompleteIncompleteStruct;
```
Comment (tritao) [inline tests/CLI/CLI.h]: What's the point of this type? It doesn't seem to be used anywhere.

**Comment by tritao** [comment]
I don't get why we need all these new checks... Why not add the new code in `FunctionIgnored` to `CheckIgnoreMethod`? Or do the checking in the pass we already have, `CheckIgnoredDecls`?

File: tests/CLI/CLI.h (hunk 13 above)
Comment (realvictorprm) [inline tests/CLI/CLI.h]: That's the point. It can't be used and if it would be compiled the result would be wrong and not useable.

Comment (realvictorprm) [inline tests/CLI/CLI.h]: So the compilation itself is the check

**Comment by ddobrev** [comment]
@tritao I thought of this myself. I decided it's better off in the CLI gen because it's about the only one which cannot deal with incomplete types.

**Pull request merged**